### PR TITLE
lsfd: print the detail of the timer associated with a timerfd 

### PR DESCRIPTION
**Pull request by masatake**
An example output:

    # ./lsfd -p 1  -Q '(TYPE == "timerfd")' -oCOMMAND,PID,ASSOC,TYPE,INODE,NAME
    COMMAND PID ASSOC    TYPE INODE NAME
    systemd   1    22 timerfd  1060 clockid=monotonic remaining=4625.661834645
    systemd   1    25 timerfd  1060 clockid=realtime remaining=398.164618943
    systemd   1   112 timerfd  1060 clockid=realtime remaining=7537606384.202715161

Signed-off-by: Masatake YAMATO <yamato@redhat.com>

### CODE REVIEW

File: misc-utils/lsfd.c [hunk 1]
```diff
@@ -294,6 +294,27 @@ static const struct colinfo infos[] = {
...
+				   0,   SCOLS_FL_RIGHT, SCOLS_JSON_STRING,
+				   N_("interval") },
+	[COL_TIMERFD_INTERVAL_INT]={"TIMERFD.INTERVAL.INT",
+				   0,   SCOLS_FL_RIGHT, SCOLS_JSON_NUMBER,
+				   N_("integer part of TIMERFD.INTERVAL") },
+	[COL_TIMERFD_INTERVAL_NANOFRAC]={"TIMERFD.INTERVAL.NANOFRAC",
+				   0,   SCOLS_FL_RIGHT, SCOLS_JSON_NUMBER,
+				   N_("the nanosecond fraction part of TIMERFD.INTERVAL") },
+	[COL_TIMERFD_REMAINING] = { "TIMERFD.REMAINING",
+				   0,   SCOLS_FL_RIGHT, SCOLS_JSON_STRING,
```
Comment (t-8ch) [inline misc-utils/lsfd.c:310]: Couldn't this be `SCOLS_JSON_NUMBER` on its own, removing the need for the two derived columns?

Comment (masatake) [reply misc-utils/lsfd.c:310]: Thank you for reviewing.

The filtering engine supports only integers. So I don't want to introduce floating numbers to lsfd.

Comment (t-8ch) [reply misc-utils/lsfd.c:310]: Wouldn't it essentially force the users to implement their own floating point logic to use this properly?

`(TIMERFD.INTERVAL.INT >= 5 && TIMERFD.INTERVAL.NANOFRAC >= 5000000) && (TIMERFD.INTERVAL.INT < 6 && TIMERFD.INTERVAL.NANOFRAC < 5000000)`

seems very unwieldy.

What would be the issues to add float support to the filtering engine?

Comment (masatake) [reply misc-utils/lsfd.c:310]: > What would be the issues to add float support to the filtering engine?

My programming skills. When I designed the filtering engine, I assumed all numbers dealt with in lsfd are integers. With this simplification, I could implement the engine.

I'll look into the code of the filtering engine to support floating point column values.

Comment (masatake) [reply misc-utils/lsfd.c]: Now we can compare floating point numbers:

```
# ./lsfd -Q '(TIMERFD.REMAINING > 1.0) && (TIMERFD.REMAINING < 30.0)'
COMMAND       PID  USER ASSOC MODE    TYPE       SOURCE MNTID INODE NAME
gsd-color   17436   jet    11  rw- timerfd anon_inodefs    15  1060 clockid=realtime remaining=6.746271218
gsd-color 3370688 test2    15  rw- timerfd anon_inodefs    15  1060 clockid=realtime remaining=6.149405906
# ./lsfd -Q '(TIMERFD.REMAINING > 0.0) && (TIMERFD.REMAINING < 1.0)' 
COMMAND          PID        USER ASSOC MODE    TYPE       SOURCE MNTID INODE NAME
Xorg           17069         jet    26  rw- timerfd anon_inodefs    15  1060 clockid=monotonic remaining=0.061075544
systemd-oomd 2382701 systemd-oom     8  rw- timerfd anon_inodefs    15  1060 clockid=monotonic remaining=0.178126915
```

File: misc-utils/lsfd-unkn.c [hunk 2]
```diff
@@ -551,6 +552,165 @@ static const struct anon_ops anon_eventpoll_ops = {
...
+			return 1;
+		}
+	}
+
+	return 0;
+}
+
+static bool is_zero_timespec(const struct timespec *t)
+{
+	return (t->tv_sec || t->tv_nsec)? false: true;
```
Comment (t-8ch) [inline misc-utils/lsfd-unkn.c:617]: `return !t->tv_sec && !t->tv_nsec;`

File: misc-utils/lsfd-unkn.c [hunk 3]
```diff
@@ -551,6 +552,165 @@ static const struct anon_ops anon_eventpoll_ops = {
...
+		data->clockid = clockid;
+		return 1;
+	} else if (strcmp(key, "it_value") == 0 ||
+		   strcmp(key, "it_interval") == 0) {
+		uint64_t tv_sec;
+		uint64_t tv_nsec;
+
+		if (sscanf(value, "(%"SCNu64", %"SCNu64")",
+			   &tv_sec, &tv_nsec) == 2) {
+			struct timespec *t = (key[3] == 'v')
```
Comment (t-8ch) [inline misc-utils/lsfd-unkn.c:603]: This check is a bit underhanded. A full strcmp would be clearer.

**Review by t-8ch — APPROVED** [review]
Thanks for incorporating the feedback!

File: misc-utils/lsfd-unkn.c (hunk 2 above)
Comment (masatake) [reply misc-utils/lsfd-unkn.c]: Done.

File: misc-utils/lsfd-unkn.c (hunk 3 above)
Comment (masatake) [reply misc-utils/lsfd-unkn.c]: To reduce calling strcmp, I changed the control flow.
The underhanded code is deleted.

**Pull request merged**